meta/src/main/scala/scalan/meta/ScalanParsers.scala

```diff
@@ -246,45 +256,89 @@ object ScalanAst {
+          throw new IllegalStateException(s"An unsupported synonym for a trait $typeDef, in module $moduleName, could not find the trait name!")
+        }
+      }
+
+
     def apply(packageName: String, imports: List[SImportStat], moduleTrait: STraitDef, config: CodegenConfig): SEntityModuleDef = {
       val moduleName = moduleTrait.name
       val defs = moduleTrait.body
 
-      val entityRepSynonym = defs.collectFirst { case t: STpeDef => t }
+      //Get all of the module-defined synonims
+      val entityRepSynonymsList = defs.collect { case t: STpeDef => t }
 
+      //The traits do not end with Companion
       val traits = defs.collect { case t: STraitDef if !t.name.endsWith("Companion") => t }
-      val entity = traits.headOption.getOrElse {
+
+      //Get all the module defined entities (UDTs)
+      val entities = defs.collect { case t: STraitDef => t }
+      entities.headOption.getOrElse {
         throw new IllegalStateException(s"Invalid syntax of entity module trait $moduleName. First member trait must define the entity, but no member traits found.")
       }
+      //Map the synonyms, if any, into the entities (UDTs)
+      var entityRepSynonyms = Map[STpeDef, STraitDef]()
+      //Iterate through all the synonyms
+      for(synonym:STpeDef <- entityRepSynonymsList){
+        //Get the trait name for a synonym
+        val traitName = getTypeTraitName(moduleName,synonym)
+        //Search for the entity (trait) with the found name
+        val entity = entities.collectFirst{case ent : STraitDef if (ent.name == traitName) => ent}.getOrElse{
+          throw new IllegalStateException(s"Invalid syntax of entity module trait $moduleName. The trait synonym is defined by the trait named $traitName is not fond.")
+        }
+        //If a trait is found then add a mapping
+        entityRepSynonyms += ( synonym -> entity )
+      }
 
-      val classes = entity.optBaseType match {
-        case Some(bt) =>
-          val entityName = entity.name
-          val entityImplName = entityName + "Impl"
-          val typeUseExprs = entity.tpeArgs.map(tpeUseExpr(_))
-          val defaultBTImpl = SClassDef(
-            name = entityImplName,
-            tpeArgs = entity.tpeArgs,
-            args = List(SClassArg(false, false, true, "wrappedValueOfBaseType", STraitCall("Rep", List(bt)), None)),
-            implicitArgs = getImplicitArgs(entity),
-            ancestors = List(STraitCall(entity.name, typeUseExprs)),
-            body = List(
-
-            ),
-            selfType = None,
-            companion = None,
-//            companion = defs.collectFirst {
-//              case c: STraitOrClassDef if c.name.toString == entityImplName + "Companion" => c
-//            },
-            true
-          )
-          defaultBTImpl :: getConcreteClasses(defs)
-        case None => getConcreteClasses(defs)
+      //Create a map of traits and their class definitions
+      var classesMap = Map[STraitDef, List[SClassDef]]()
```

Comment:
No need for a `var` here:

    val classesMap = entities.map(entity => entity -> entity -> getEntityClasses(entity, defs)).toMap